tests/test_contracts/test_base.py

```diff
@@ -199,6 +207,78 @@ def test_scaffold():
         scaffold.get_state("ledger_api", "contract_address", **kwargs)
 
 
+@mock.patch("aea.cli.utils.decorators._cast_ctx")
+def test_scaffolded_contract_method_call():
+    """Tests a contract method call."""
+
+    # Mock the CLI context
+    td = mkdtemp()
+
+    ctx = ContextMock()
+    ctx.agent_config.author = "dummy_author"
+    ctx.config["to_local_registry"] = True
+    ctx.agent_config.directory = td
+
+    contract_name = "IUniswapV2ERC20"
+    contract_abi = "./IUniswapV2ERC20.json"
+
+    try:
+        # Scaffold a new contract
+        scaffold_item(ctx, CONTRACT, contract_name)
+        add_contract_abi(ctx, contract_name, contract_abi)
+
+        # Load the new contract
+        contract = Contract.from_dir(td)
+        ledger_api = ledger_apis_registry.make(
+            EthereumCrypto.identifier,
+            address=ETHEREUM_DEFAULT_ADDRESS,
+        )
+
+        # Call a contract method: allowance
+        CHAIN_ID = 1
+        CONTRACT_ADDRESS = "0xa6B71E26C5e0845f74c812102Ca7114b6a896AB2"
+        SPENDER_ADDRESS = "0x7A1236d5195e31f1F573AD618b2b6FEFC85C5Ce6"
+        SENDER_ADDRESS = "0x7A1236d5195e31f1F573AD618b2b6FEFC85C5Ce6"
+        APPROVAL_VALUE = 100
+        GAS = 100
+        GAS_PRICE = 100
+        NONCE = 0
+        VALUE = 0
+
+        res = contract.build_transaction(
+            ledger_api,
+            "allowance",
+            method_args={
+                "spender": SPENDER_ADDRESS,
+                "value": APPROVAL_VALUE,
+            },
+            tx_args={
+                "sender_address": SENDER_ADDRESS,
+                "gas": GAS,
+                "gasPrice": GAS_PRICE,
+            },
+        )
+
+        data = contract.get_instance(ledger_api, CONTRACT_ADDRESS).encodeABI(
+            fn_name="approve", args=[SPENDER_ADDRESS, APPROVAL_VALUE]
+        )
```

Comment:
Ok that's part of the way of what I want to see. But now next step: add functionality and test demonstrating its usage so that a message from skill can be sent to contract and interpreted by ledger connection without any custom method needing implementing on the contract. So the message is interpreted by ledger connection, handed to a "default" method on the contract and processed there, then returned values passed back into a message and sent back to skill. It's enough to demo this on the ledger connection and contract side only, no need for a skill.

Reply:
> So the message is interpreted by ledger connection, handed to a "default" method (which automatically detects what method on abi to call based on args, kwargs) on the contract

Not sure I completely follow: why do we need a default method? Isn't it enough that a dev calls a method using either build_transaction for "write" methods or contract_method_call for "read" ones? Why would we be guessing the method? A dev should provide the method name.

Reply:
The whole point is we can have a no-code solution here; where the base contract directly handles the messages coming from the skills and interpreted in ledger connection. You need to understand ledger connection to be able to understand this suggestion.